packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -31,19 +33,164 @@ import PropTypes from 'prop-types'
+    })
+    const [queue, setQueue] = useState({})
+
+    // iterates through each deep nested object and if finds object that has prop and value specified in objToFindBy
+    // argument, it replaces the current object with replacementObj, stops recursive walk and returns the whole tree.
+    // If none is found, it returns false.
+
+    // copied from https://github.com/brojd/obj-traverse/blob/master/src/findAndModifyFirst/findAndModifyFirst.js
+    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
+        let treeToReturn = tree
+        let findSuccess = false
+        let modifiedObj = false
+        const findKeys = Object.keys(objToFindBy)
+        findKeys.forEach((key) => {
+            isEqual(tree[key], objToFindBy[key]) ? (findSuccess = true) : (findSuccess = false)
+        })
+        if (findSuccess) {
+            for (let prop in tree) {
+                delete tree[prop]
+            }
+            for (let prop in replacementObj) {
+                tree[prop] = replacementObj[prop]
+            }
+            return tree
+        }
+        const findInChildren = (obj, childrenKey, objToFindBy, replacementObj) => {
+            if (Object.prototype.hasOwnProperty.call(obj, childrenKey)) {
+                for (let i = 0; i < obj[childrenKey].length; i++) {
+                    findKeys.forEach((key) => {
+                        isEqual(obj[childrenKey][i][key], objToFindBy[key])
+                            ? (findSuccess = true)
+                            : (findSuccess = false)
+                    })
+                    if (findSuccess) {
+                        obj[childrenKey][i] = replacementObj
+                        modifiedObj = true
+                        break
+                    }
+                }
+                if (!findSuccess) {
+                    obj[childrenKey].forEach((child) =>
+                        findInChildren(child, childrenKey, objToFindBy, replacementObj)
+                    )
+                }
+            }
+            return obj
+        }
+        findInChildren(tree, childrenKey, objToFindBy, replacementObj)
+        return modifiedObj ? treeToReturn : false
+    }
+
+    const fetchCategoryNode = async (id, depth = 1) => {
+        const STALE_TIME = 10000
+        let res
+        // return early if there's an in-flight request or one that is less
+        // than the stale time
+        if (queue[id] === 'loading' || Date.now() < queue[id] + STALE_TIME) {
+            return
+        }
+        setQueue({
+            ...queue,
+            [id]: 'loading'
+        })
+        if (!testing) {
+            res = await api.shopperProducts.getCategory({
+                parameters: {
+                    id,
+                    levels: depth
+                }
+            })
+        } else {
+            res = {
+                id: 'mens-clothing',
+                name: 'Clothing',
+                loaded: true
+            }
+        }
```

Comment:
This approach is a maintenance risk because we've got mock data code buried in our application code. We are mixing concerns (one concern is the application, the other is the testing of the application). 

If we have no other option but this approach, then you would want to do something like this instead so that the mock data does not appear in any application code and only gets declared in testing code (e.g. the mock providers test util helper)

```<CategoriesProvider testData={{
                id: 'mens-clothing',
                name: 'Clothing',
                loaded: true
            }}> ... </CategoriesProvider>
```